includes/plugins/class-woocommerce-memberships.php

```diff
@@ -0,0 +1,195 @@
+				}
+
+				$is_post_restricted = wc_memberships_is_post_content_restricted( $list_object->get_id() );
+
+				if ( ! $is_post_restricted ) {
+					return true;
+				}
+
+				$user_id = self::$user_id_in_scope ?? get_current_user_id();
+
+				return wc_memberships_user_can( $user_id, 'view', [ 'post' => $list_object->get_id() ] );
+
+			} 
+		);
+		return array_values( $lists );
+	}
+
+	/**
+	 * Receives an array of Lists and returns only the ones that the user has access to
+	 *
+	 * @param array $lists An array of lists in which the key are the list IDs.
+	 * @return array
+	 */
+	public static function filter_lists_objects( $lists ) {
+		if ( ! is_array( $lists ) || empty( $lists ) ) {
+			return $lists;
+		}
+
+		$list_ids = self::filter_lists( array_keys( $lists ) );
+
+		return array_filter(
+			$lists,
+			function( $list_id ) use ( $list_ids ) {
+				return in_array( $list_id, $list_ids );
+			},
+			ARRAY_FILTER_USE_KEY
+		);
+
+	}
+
+	/**
+	 * Remove lists that require a membership plan when the membership is cancelled
+	 *
+	 * @param WC_Memberships_User_Membership $user_membership The User Membership object.
+	 * @param string                         $old_status old status, without the `wcm-` prefix.
+	 * @param string                         $new_status new status, without the `wcm-` prefix.
+	 * @return void
+	 */
+	public static function remove_user_from_list( $user_membership, $old_status, $new_status ) {
+		$status_considered_active = wc_memberships()->get_user_memberships_instance()->get_active_access_membership_statuses();
+
+		if ( in_array( $new_status, $status_considered_active ) ) {
+			return;
+		}
+		
+		$lists_to_remove = [];
+		$user            = $user_membership->get_user();
+		if ( ! $user ) {
+			return;
+		}
+		$user_email = $user->user_email;
+		$rules      = $user_membership->get_plan()->get_content_restriction_rules();
+		foreach ( $rules as $rule ) {
+			if ( Subscription_Lists::CPT !== $rule->get_content_type_name() ) {
+				continue;
+			}
+			$object_ids = $rule->get_object_ids();
+			foreach ( $object_ids as $object_id ) {
+				$subscription_list = new Subscription_List( $object_id );
+				$lists_to_remove[] = $subscription_list->get_form_id();
+			}
+		}
+		$provider = Newspack_Newsletters::get_service_provider();
+		$provider->update_contact_lists_handling_local( $user_email, [], $lists_to_remove );
+		
+	}
+
+	/**
+	 * Adds user to premium lists when a membership is granted
+	 *
+	 * @param \WC_Memberships_Membership_Plan $plan the plan that user was granted access to.
+	 * @param array                           $args 
+	 * {
+	 *     Array of User Membership arguments.
+	 * 
+	 *     @type int $user_id the user ID the membership is assigned to.
+	 *     @type int $user_membership_id the user membership ID being saved.
+	 *     @type bool $is_update whether this is a post update or a newly created membership.
+	 * }
+	 * @return void
+	 */
+	public static function add_user_to_list( $plan, $args ) {
```

Comment:
This method did not work as expected for me.

I purchased the subscription which grants me the membership. The method here is reached but the user never gets subscribed to the newsletters.

While testing I also noticed that `wc_memberships_user_membership_created` is not called when a membership is manually granted through the "Members" admin dashboard.

Reply:
@miguelpeixe 

> I purchased the subscription which grants me the membership. The method here is reached but the user never gets subscribed to the newsletters.

Can you double-check this? Or detail the steps you took? I can't reproduce it. This has been tested multiple times with different ESPs, in different sites including some "real" sites and it works every time

> While testing I also noticed that wc_memberships_user_membership_created is not called when a membership is manually granted through the "Members" admin dashboard.

I'm gonna check this

Reply:
@miguelpeixe https://github.com/Automattic/newspack-newsletters/pull/1238/commits/69f825bf309907cf9ff4c61da233bd72419b988e makes it work when adding a membership via admin. It also removes the user if the membership is deleted (before this, it would only remove when the membership was cancelled or paused, but not deleted)

The `wc_memberships_user_membership_saved` I'm using now is a bit inconsistent and hard to rely on. It may be called multiple times during the same action, its parameters are not always what the documentation says they will be, so I kept it only for adding subscription, but I'm still relying on other hook for removing subscriptions.

Note that there's something we can improve but that's not covered by this first phase of the project:

Imagine someone buys a membership, then unsubscribe to a premium Newsletter. Later, their subscription expires and our system tries to remove them for the Premium list. Fine.

Then they renew their subscription. When this happens, they will be re-subscribed to the list, even though they have unsubscribed before.

We can keep a record and not do it in the future.

--

Please test all the cases
- membership behind registration and product
- acquiring membership via registration or buying a product
- manually adding members
- manually canceling
- manually deleting memberships
- manually changing statuses...
- etc...

Reply:
I do find the idea of subscribing someone to a newsletter list automatically when getting a membership a bit non-intuitive from a technical perspective. We may find ourselves decoupling that logic in the future so it's not a default behavior.

Reply:
That is the current request from a business that fits into the MVP.

In the future I think we'll want to have an opt in in the forms and let the users choose